llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp

```diff
@@ -1991,9 +2068,155 @@ void FunctionEncoder::handleMemoryPhiNode(MemoryPhi &mphi, int passID) {
+    throw std::runtime_error(
+        "[handleIntrinsicAddWithOverflow]"
+        "Unsupported type for add_with_overflow intrinsic."
+        "Only struct types like {i32, i1} or {i64, i1} are supported\n");
+  }
+  outs() << "[handleIntrinsicAddWithOverflow]"
+         << "instRetType: " << *instRetType << "\n";
+
+  ValueBVTreeMap *currentBBValueBVTreeMap;
+  if (this->BBValueBVTreeMap.find(currentBB) == this->BBValueBVTreeMap.end()) {
+    currentBBValueBVTreeMap = new ValueBVTreeMap();
+    BBValueBVTreeMap.insert({currentBB, currentBBValueBVTreeMap});
+  } else {
+    currentBBValueBVTreeMap = this->BBValueBVTreeMap.at(currentBB);
+  }
+
+  // the call instruction returns a struct of type {i32, i1},
+  // create a bitvector tree of the same structure.
+  BVTree *newTree = setupBVTreeForArg(intrinsicInstValue,
+                                      intrinsicInstValue->getName().str());
+  outs() << "[handleIntrinsicAddWithOverflow] "
+         << "newTree: " << newTree->toString() << "\n";
+  currentBBValueBVTreeMap->insert({intrinsicInstValue, newTree});
+
+  auto resBV = newTree->getSubTree(0)->bv;
+  auto overflowBV = newTree->getSubTree(1)->bv;
+  outs() << "[handleIntrinsicAddWithOverflow] "
+         << "resBV: " << resBV.get_sort().to_string().c_str() << "\n";
+  outs() << "[handleIntrinsicAddWithOverflow] "
+         << "overflowBV: " << overflowBV.get_sort().to_string().c_str() << "\n";
+
+  z3::expr resultExprNoOverflowOrUnderflow(ctx), resultExpr(ctx);
+  if (intrinsicInstID == Intrinsic::sadd_with_overflow) {
+    resultExprNoOverflowOrUnderflow =
+        z3::ite((z3::bvadd_no_overflow(callOperand0BV, callOperand1BV, true) &&
+                 z3::bvadd_no_underflow(callOperand0BV, callOperand1BV)),
+                overflowBV == 0, overflowBV == 1);
+    resultExpr = resBV == callOperand0BV + callOperand1BV;
+  } else if (intrinsicInstID == Intrinsic::uadd_with_overflow) {
+    resultExprNoOverflowOrUnderflow =
+        z3::ite((z3::bvadd_no_overflow(callOperand0BV, callOperand1BV, false) &&
+                 z3::bvadd_no_underflow(callOperand0BV, callOperand1BV)),
+                overflowBV == 0, overflowBV == 1);
+    resultExpr = resBV == callOperand0BV + callOperand1BV;
+  } else if (intrinsicInstID == Intrinsic::ssub_with_overflow) {
+    resultExprNoOverflowOrUnderflow =
+        z3::ite((z3::bvsub_no_overflow(callOperand0BV, callOperand1BV) &&
+                 z3::bvsub_no_underflow(callOperand0BV, callOperand1BV, true)),
+                overflowBV == 0, overflowBV == 1);
+    resultExpr = resBV == callOperand0BV - callOperand1BV;
+  } else if (intrinsicInstID == Intrinsic::usub_with_overflow) {
+    resultExprNoOverflowOrUnderflow =
+        z3::ite((z3::bvsub_no_overflow(callOperand0BV, callOperand1BV) &&
+                 z3::bvsub_no_underflow(callOperand0BV, callOperand1BV, false)),
+                overflowBV == 0, overflowBV == 1);
+    resultExpr = resBV == callOperand0BV - callOperand1BV;
+  }
```

Comment:
Could we do a `switch (intrinsicInstID)` instead of `if/else` and error in the `default` case (instead of having to do it in the caller)?

Reply:
It's a good idea to use switch statements. 

I thought it makes sense to keep the check in the caller, because this particular function `handleAddSubWithOverflowIntrinsic` is meant to handle only the four intrinsics add/sub/overflow/underflow.  So any function that calls it should make sure it is one of those.